Fixed problem with line feed in branch name

jonmills opened the pull request:
My hg repo included a repo with a line feed in the branch name, which caused hg-fast-export.py to crash. I've added a fix - although my regex skills are not good enough to make it fit neatly with your solution.

frej commented:
I don't like these kind of ad-hoc fixes, what should happen if your "fix" turns the branch name into a name that exists?

The way to do it properly is to correct the branch name using a branch mapping file (`-B`). Unfortunately the current release has no support for encoding `\r` in the mapping. This has irritated my for a while and your patch triggered me to do it properly. Try the branch [new-mapping-format,](https://github.com/frej/fast-export/tree/new-mapping-format) here the mapping file format is changed to `"<escaped-string>"="<escaped-string>"` so you should, for example, be able to say:

`"broken\rbranch"="broken-branch"`

Please give it a spin and report back, if it works it will go in the next release.

jonmills commented:
Thanks for you help on this. I tried the new-mapping-format branch with the following branch mapping file contents (to test the operation on the branch with the line feed) :

`"CommLoss_\nPAKMODSERV-84"="CommLoss_PAKMODSERV-84"`

and got the following error:

`
master: Exporting full revision 1/10215 with 9/0/0 added/changed/removed files
fatal: Unsupported command: <_sre.SRE_Match object at 0x7fcfbba47940>
fast-import: dumping crash report to .git/fast_import_crash_12104
Traceback (most recent call last):
  File "/externals/fast-export/hg-fast-export.py", line 526, in <module>
    notes=options.notes,encoding=encoding,fn_encoding=fn_encoding))
  File "/externals/fast-export/hg-fast-export.py", line 424, in hg2git
    sob,brmap,hgtags,encoding,fn_encoding)
  File "/externals/fast-export/hg-fast-export.py", line 244, in export_commit
    export_file_contents(ctx,man,added,hgtags,fn_encoding)
  File "/externals/fast-export/hg-fast-export.py", line 141, in export_file_contents
    wr(d)
  File "/externals/fast-export/hg-fast-export.py", line 37, in wr
    wr_no_nl(msg)
  File "/externals/fast-export/hg-fast-export.py", line 34, in wr_no_nl
    sys.stdout.write(msg)
IOError: [Errno 32] Broken pipe
fatal: You are on a branch yet to be born
`

So I tried just mapping one of the normal branches:
`"Sprint 14 Completed"="Random_test_branch"`

and got a similar error:

`
master: Exporting full revision 1/10215 with 9/0/0 added/changed/removed files
fatal: Unsupported command: <_sre.SRE_Match object at 0x7ff4fcdfe940>
fast-import: dumping crash report to .git/fast_import_crash_27824
Traceback (most recent call last):
  File "/externals/fast-export/hg-fast-export.py", line 526, in <module>
    notes=options.notes,encoding=encoding,fn_encoding=fn_encoding))
  File "/externals/fast-export/hg-fast-export.py", line 424, in hg2git
    sob,brmap,hgtags,encoding,fn_encoding)
  File "/externals/fast-export/hg-fast-export.py", line 244, in export_commit
    export_file_contents(ctx,man,added,hgtags,fn_encoding)
  File "/externals/fast-export/hg-fast-export.py", line 141, in export_file_contents
    wr(d)
  File "/externals/fast-export/hg-fast-export.py", line 37, in wr
    wr_no_nl(msg)
  File "/externals/fast-export/hg-fast-export.py", line 34, in wr_no_nl
    sys.stdout.write(msg)
IOError: [Errno 32] Broken pipe
fatal: You are on a branch yet to be born
`

I also got the same error when running with a branch map with a single non-existent branch name in it:

`"Non existent branch"="Non_existent_branch"`

I didn't see this error when I ran with no branch map file.

frej commented:
Serves me right for not testing properly, there was a forgotten trace
printout in there. Please try the refreshed branch.

jonmills commented:
Thanks for the quick response. That works great.

frej commented:
Workaround is available in v171002, closing.
